src/lazy/binary/raw/v1_1/immutable_buffer.rs

```diff
@@ -275,76 +277,198 @@ impl<'a> ImmutableBuffer<'a> {
+        if opcode.is_nop() {
             // ...skip through NOPs until we found the next non-NOP byte.
-            (_, input) = self.consume_nop_padding(type_descriptor)?;
+            (_, input) = self.consume_nop_padding(opcode)?;
             // If there is no next byte, we're out of values.
             if input.is_empty() {
                 return Ok(None);
             }
             // Otherwise, there's a value.
-            type_descriptor = input.peek_opcode()?;
+            opcode = input.peek_opcode()?;
         }
-        if type_descriptor.is_e_expression() {
+        if opcode.is_e_expression() {
             return Ok(Some(RawValueExpr::EExp(
-                self.read_e_expression(type_descriptor)?,
+                self.read_e_expression(opcode)?,
             )));
         }
         Ok(Some(RawValueExpr::ValueLiteral(
-            input.read_value(type_descriptor)?,
+            input.read_value(opcode)?,
         )))
     }
 
-    /// Reads a value from the buffer. The caller must confirm that the buffer is not empty and that
-    /// the next byte (`type_descriptor`) is not a NOP.
-    pub fn read_value(self, opcode: Opcode) -> IonResult<LazyRawBinaryValue_1_1<'a>> {
-        if opcode.is_annotations_sequence() {
-            self.read_annotated_value(opcode)
+    pub(crate) fn peek_delimited_container(
+        self,
+        opcode: Opcode,
+    ) -> IonResult<LazyRawBinaryValue_1_1<'a>> {
+        use crate::IonType;
+
+        if let Some(IonType::Struct) = opcode.ion_type {
+            self.peek_delimited_struct()
         } else {
-            self.read_value_without_annotations(opcode)
+            self.peek_delimited_sequence()
         }
     }
 
-    /// Reads a value from the buffer. The caller must confirm that the buffer is not empty and that
-    /// the next byte (`type_descriptor`) is neither a NOP nor an annotations wrapper.
-    fn read_value_without_annotations(
-        self,
-        type_descriptor: Opcode,
-    ) -> IonResult<LazyRawBinaryValue_1_1<'a>> {
-        let input = self;
-        let header = type_descriptor
+    pub(crate) fn peek_delimited_sequence(self) -> IonResult<LazyRawBinaryValue_1_1<'a>> {
+        let head_opcode = self.peek_opcode()?;
+
+        let mut input = self.consume(1);
+        let header_offset = input.offset();
+        let mut offsets = BumpVec::<usize>::new_in(self.context.allocator());
+
+        loop {
+            let opcode = input.peek_opcode()?;
+            if opcode.opcode_type == OpcodeType::DelimitedContainerClose {
+                offsets.push(input.offset());
+                break;
+            } else if opcode.opcode_type == OpcodeType::Nop {
+                let res = input.consume_nop_padding(opcode)?;
+                input = res.1;
+            } else if let Some(value) = ImmutableBuffer::peek_sequence_value_expr(input)? {
+                offsets.push(input.offset());
+                input = input.consume(value.range().len());
+            }
```

Comment:
`peek_sequence_value_expr(input)` is fully lexing the value, is there a reason you decided to store the offset rather than the `LazyRawValueExpr` (`value`) itself?

Reply:
I had a lifetime issue with the values when I first started working on it. Now that I think about it though, I had some mismatched lifetimes that I ended up fixing later that could have been the problem. I'l revisit.